File: eip_toolkit_simple.py
#!/usr/bin/env python3
"""
EIP Toolkit - Simple Version
Monitor Azure Red Hat OpenShift Egress IP assignments and CPIC status.
"""

import argparse
import json
import logging
import os
import subprocess
import sys
import time
from datetime import datetime
from pathlib import Path

import matplotlib.pyplot as plt
import pandas as pd


class EIPToolkitError(Exception):
    """Custom exception for EIP Toolkit errors."""
    pass


class EIPMonitor:
    """Simple EIP monitoring class."""
    
    def __init__(self, output_dir: Path):
        self.output_dir = output_dir
        self.logs_dir = output_dir / 'logs'
        self.data_dir = output_dir / 'data'
        self.plots_dir = output_dir / 'plots'
        
        for directory in [self.logs_dir, self.data_dir, self.plots_dir]:
            directory.mkdir(parents=True, exist_ok=True)
        
        logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
        self.logger = logging.getLogger(__name__)
    
    def run_oc_command(self, cmd):
        """Run oc command and return JSON result."""
        try:
            result = subprocess.run(['oc'] + cmd, capture_output=True, text=True, check=True)
            return json.loads(result.stdout)
        except (subprocess.CalledProcessError, json.JSONDecodeError) as e:
            raise EIPToolkitError(f"OpenShift command failed: {e}")
    
    def run_az_command(self, cmd):
        """Run az command and return JSON result."""
        try:
            subscription = os.getenv('AZ_SUBSCRIPTION')
            if not subscription:
                raise EIPToolkitError("AZ_SUBSCRIPTION environment variable not set")
            
            result = subprocess.run(['az'] + cmd + ['--subscription', subscription], 
                                  capture_output=True, text=True, check=True)
            return json.loads(result.stdout)
        except (subprocess.CalledProcessError, json.JSONDecodeError) as e:
            raise EIPToolkitError(f"Azure command failed: {e}")
    
    def get_eip_stats(self):
        """Get EIP statistics."""
        data = self.run_oc_command(['get', 'eip', '-o', 'json'])
        assigned = len([item for item in data['items'] if item.get('status', {}).get('items', [])])
        unassigned = len([item for item in data['items'] if not item.get('status', {}).get('items', [])])
        configured = len(data['items'])
        return {'configured': configured, 'assigned': assigned, 'unassigned': unassigned}
    
    def get_cpic_stats(self):
        """Get CPIC statistics."""
        data = self.run_oc_command(['get', 'cloudprivateipconfig', '-o', 'json'])
        success = len([item for item in data['items'] 
                      for condition in item.get('status', {}).get('conditions', [])
                      if condition.get('reason') == 'CloudResponseSuccess'])
        pending = len([item for item in data['items'] 
                      for condition in item.get('status', {}).get('conditions', [])
                      if condition.get('reason') == 'CloudResponsePending'])
        error = len([item for item in data['items'] 
                    for condition in item.get('status', {}).get('conditions', [])
                    if condition.get('reason') == 'CloudResponseError'])
        return {'success': success, 'pending': pending, 'error': error}
    
    def get_nodes(self):
        """Get EIP-enabled nodes."""
        try:
            result = subprocess.run(['oc', 'get', 'nodes', '-l', 'k8s.ovn.org/egress-assignable=true', '-o', 'name'],
                                  capture_output=True, text=True, check=True)
            return [line.replace('node/', '') for line in result.stdout.strip().split('\n') if line]
        except subprocess.CalledProcessError as e:
            raise EIPToolkitError(f"Cannot access OpenShift nodes: {e.stderr}")
    
    def log_stats(self, timestamp, stats_type, **kwargs):
        """Log statistics to files."""
        for stat_name, value in kwargs.items():
            log_file = self.logs_dir / f"{stats_type}_{stat_name}.log"
            with open(log_file, 'a') as f:
                f.write(f"{timestamp} {value}\n")
    
    def monitor(self):
        """Main monitoring loop."""
        self.logger.info("Starting EIP monitoring...")
        
        nodes = self.get_nodes()
        self.logger.info(f"Found EIP-enabled nodes: {nodes}")
        
        while True:
            timestamp = datetime.now().isoformat()
            
            # Get global statistics
            eip_stats = self.get_eip_stats()
            cpic_stats = self.get_cpic_stats()
            
            # Log global statistics
            self.log_stats(timestamp, 'ocp_eips', **eip_stats)
            self.log_stats(timestamp, 'ocp_cpic', **cpic_stats)
            
            # Monitor individual nodes
            for node in nodes:
                try:
                    # Get node-specific EIP stats
                    eip_data = self.run_oc_command(['get', 'eip', '-o', 'json'])
                    eip_assigned = len([item for item in eip_data['items']
                                      for status_item in item.get('status', {}).get('items', [])
                                      if status_item.get('node') == node])
                    
                    # Get node-specific CPIC stats
                    cpic_data = self.run_oc_command(['get', 'cloudprivateipconfig', '-o', 'json'])
                    cpic_success = len([item for item in cpic_data['items']
                                      if (item.get('spec', {}).get('node') == node and
                                          any(c.get('reason') == 'CloudResponseSuccess'
                                              for c in item.get('status', {}).get('conditions', [])))])
                    
                    # Log node statistics
                    self.log_stats(timestamp, f'{node}_ocp_eip', assigned=eip_assigned)
                    self.log_stats(timestamp, f'{node}_ocp_cpic', success=cpic_success)
                    
                    self.logger.info(f"{node} - EIP: {eip_assigned}, CPIC: {cpic_success}")
                    
                except Exception as e:
                    self.logger.error(f"Error monitoring node {node}: {e}")
            
            # Check if monitoring should continue
            if (eip_stats['assigned'] == eip_stats['configured'] and 
                cpic_stats['success'] == eip_stats['configured']):
                break
            
            time.sleep(1)
        
        self.logger.info("Monitoring complete - all EIPs assigned and CPIC issues resolved")
    
    def merge_logs(self):
        """Merge log files into data files."""
        self.logger.info("Merging log files...")
        
        # Get all log files
        log_files = list(self.logs_dir.glob('*.log'))
        if not log_files:
            raise EIPToolkitError("No log files found")
        
        # Get unique node names
        nodes = set()
        for log_file in log_files:
            filename = log_file.name
            if not filename.startswith(('ocp_', 'azure_')) and '_' in filename:
                node_name = filename.split('_')[0]
                nodes.add(node_name)
        
        # Process each file type
        file_mappings = {
            'ocp_cpic_success.log': 'ocp_cpic_success.dat',
            'ocp_cpic_pending.log': 'ocp_cpic_pending.dat', 
            'ocp_cpic_error.log': 'ocp_cpic_error.dat',
            'ocp_eip_assigned.log': 'ocp_eip_assigned.dat'
        }
        
        for log_suffix, dat_filename in file_mappings.items():
            data_file = self.data_dir / dat_filename
            
            with open(data_file, 'w') as outfile:
                for node in sorted(nodes):
                    log_file = self.logs_dir / f"{node}_{log_suffix}"
                    if log_file.exists():
                        outfile.write(f'"{node}"\n')
                        with open(log_file, 'r') as infile:
                            outfile.write(infile.read())
                        outfile.write('\n\n')
        
        self.logger.info("Log merge completed")
    
    def create_plots(self):
        """Generate plots from data files."""
        self.logger.info("Creating plots...")
        
        data_files = list(self.data_dir.glob('*.dat'))
        if not data_files:
            raise EIPToolkitError("No .dat files found for plotting")
        
        for data_file in data_files:
            try:
                # Read the data file
                with open(data_file, 'r') as f:
                    content = f.read()
                
                # Parse data sections
                sections = content.split('\n\n')
                
                plt.figure(figsize=(12, 8))
                colors = ['green', 'red', 'blue', 'violet', 'orange', 'brown']
                
                for i, section in enumerate(sections):
                    if not section.strip():
                        continue
                    
                    lines = section.strip().split('\n')
                    if not lines:
                        continue
                    
                    node_name = lines[0].strip('"')
                    data_lines = lines[1:]
                    
                    timestamps = []
                    values = []
                    
                    for line in data_lines:
                        if line.strip():
                            parts = line.split()
                            if len(parts) >= 2:
                                try:
                                    timestamp = pd.to_datetime(parts[0])
                                    value = float(parts[1])
                                    timestamps.append(timestamp)
                                    values.append(value)
                                except (ValueError, pd.errors.ParserError):
                                    continue
                    
                    if timestamps and values:
                        color = colors[i % len(colors)]
                        plt.plot(timestamps, values, label=node_name, 
                                linewidth=2, color=color, marker='o', markersize=4)
                
                # Format plot
                title = data_file.stem.replace('_', ' ').title()
                plt.title(title, fontsize=14, fontweight='bold')
                plt.xlabel('Time', fontsize=12)
                plt.ylabel('Value', fontsize=12)
                plt.legend()
                plt.grid(True, alpha=0.3)
                plt.xticks(rotation=45)
                plt.tight_layout()
                
                # Save plot
                plot_file = self.plots_dir / f"{data_file.stem}.png"
                plt.savefig(plot_file, dpi=150, bbox_inches='tight')
                plt.close()
                
                self.logger.info(f"Generated plot: {plot_file}")
                
            except Exception as e:
                self.logger.error(f"Error creating plot for {data_file}: {e}")
        
        self.logger.info("Plot generation completed")


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description='EIP Toolkit - Simple Version')
    subparsers = parser.add_subparsers(dest='command', help='Available commands')
    
    # Monitor command
    subparsers.add_parser('monitor', help='Monitor EIP and CPIC status')
    
    # Merge command
    merge_parser = subparsers.add_parser('merge', help='Merge log files into data files')
    merge_parser.add_argument('directory', help='Directory containing log files')
    
    # Plot command
    plot_parser = subparsers.add_parser('plot', help='Generate plots from data files')
    plot_parser.add_argument('directory', help='Directory containing data files')
    
    # All command
    subparsers.add_parser('all', help='Run complete pipeline: monitor → merge → plot')
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return 1
    
    try:
        if args.command == 'monitor':
            timestamp = datetime.now().strftime('%y%m%d_%H%M%S')
            output_dir = Path('../runs') / timestamp
            monitor = EIPMonitor(output_dir)
            monitor.monitor()
            
        elif args.command == 'merge':
            base_dir = Path(args.directory)
            if not base_dir.exists():
                raise EIPToolkitError(f"Directory {base_dir} does not exist")
            monitor = EIPMonitor(base_dir)
            monitor.merge_logs()
            
        elif args.command == 'plot':
            base_dir = Path(args.directory)
            if not base_dir.exists():
                raise EIPToolkitError(f"Directory {base_dir} does not exist")
            monitor = EIPMonitor(base_dir)
            monitor.create_plots()
            
        elif args.command == 'all':
            timestamp = datetime.now().strftime('%y%m%d_%H%M%S')
            output_dir = Path('../runs') / timestamp
            monitor = EIPMonitor(output_dir)
            
            print("🚀 Starting Complete EIP Pipeline")
            print("📊 Phase 1: Monitoring...")
            monitor.monitor()
            print("✅ Phase 1 Complete")
            
            print("🔄 Phase 2: Merging logs...")
            monitor.merge_logs()
            print("✅ Phase 2 Complete")
            
            print("📈 Phase 3: Creating plots...")
            monitor.create_plots()
            print("✅ Phase 3 Complete")
            
            print("🎉 PIPELINE COMPLETE!")
            print(f"📁 Outputs saved in: {output_dir}")
        
        return 0
        
    except EIPToolkitError as e:
        print(f"Error: {e}")
        return 1
    except KeyboardInterrupt:
        print("Interrupted by user")
        return 1
    except Exception as e:
        print(f"Unexpected error: {e}")
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: eip_toolkit_simple.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

var (
	outputDir string
	logsDir   string
	dataDir   string
	plotsDir  string
)

// EIPToolkitError represents toolkit-specific errors
type EIPToolkitError struct {
	Message string
}

func (e *EIPToolkitError) Error() string {
	return e.Message
}

// setupDirectories creates necessary output directories
func setupDirectories(baseDir string) error {
	logsDir = filepath.Join(baseDir, "logs")
	dataDir = filepath.Join(baseDir, "data")
	plotsDir = filepath.Join(baseDir, "plots")

	for _, dir := range []string{logsDir, dataDir, plotsDir} {
		if err := os.MkdirAll(dir, 0755); err != nil {
			return fmt.Errorf("failed to create directory %s: %w", dir, err)
		}
	}

	return nil
}

// runOCCommand executes oc command and returns JSON result
func runOCCommand(cmd []string) (map[string]interface{}, error) {
	args := append([]string{}, cmd...)
	cmdObj := exec.Command("oc", args...)
	output, err := cmdObj.Output()
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			return nil, fmt.Errorf("OpenShift command failed: %s", string(exitError.Stderr))
		}
		return nil, fmt.Errorf("OpenShift command failed: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	return result, nil
}

// runAZCommand executes az command and returns JSON result
func runAZCommand(cmd []string) (interface{}, error) {
	subscription := os.Getenv("AZ_SUBSCRIPTION")
	if subscription == "" {
		return nil, &EIPToolkitError{Message: "AZ_SUBSCRIPTION environment variable not set"}
	}

	args := append([]string{}, cmd...)
	args = append(args, "--subscription", subscription)
	cmdObj := exec.Command("az", args...)
	output, err := cmdObj.Output()
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			return nil, fmt.Errorf("Azure command failed: %s", string(exitError.Stderr))
		}
		return nil, fmt.Errorf("Azure command failed: %w", err)
	}

	var result interface{}
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	return result, nil
}

// getEIPStats returns EIP statistics
func getEIPStats() (map[string]int, error) {
	data, err := runOCCommand([]string{"get", "eip", "-o", "json"})
	if err != nil {
		return nil, err
	}

	items, ok := data["items"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid items format")
	}

	configured := len(items)
	assigned := 0
	unassigned := 0

	for _, item := range items {
		itemMap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		status, ok := itemMap["status"].(map[string]interface{})
		if !ok {
			unassigned++
			continue
		}

		statusItems, ok := status["items"].([]interface{})
		if ok && len(statusItems) > 0 {
			assigned++
		} else {
			unassigned++
		}
	}

	return map[string]int{
		"configured": configured,
		"assigned":   assigned,
		"unassigned": unassigned,
	}, nil
}

// getCPICStats returns CPIC statistics
func getCPICStats() (map[string]int, error) {
	data, err := runOCCommand([]string{"get", "cloudprivateipconfig", "-o", "json"})
	if err != nil {
		return nil, err
	}

	items, ok := data["items"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid items format")
	}

	success := 0
	pending := 0
	errorCount := 0

	for _, item := range items {
		itemMap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		status, ok := itemMap["status"].(map[string]interface{})
		if !ok {
			continue
		}

		conditions, ok := status["conditions"].([]interface{})
		if !ok {
			continue
		}

		for _, cond := range conditions {
			condMap, ok := cond.(map[string]interface{})
			if !ok {
				continue
			}

			reason, ok := condMap["reason"].(string)
			if !ok {
				continue
			}

			switch reason {
			case "CloudResponseSuccess":
				success++
			case "CloudResponsePending":
				pending++
			case "CloudResponseError":
				errorCount++
			}
		}
	}

	return map[string]int{
		"success": success,
		"pending": pending,
		"error":   errorCount,
	}, nil
}

// getNodes returns list of EIP-enabled nodes
func getNodes() ([]string, error) {
	cmd := exec.Command("oc", "get", "nodes", "-l", "k8s.ovn.org/egress-assignable=true", "-o", "name")
	output, err := cmd.Output()
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			return nil, &EIPToolkitError{Message: fmt.Sprintf("Cannot access OpenShift nodes: %s", string(exitError.Stderr))}
		}
		return nil, fmt.Errorf("cannot access OpenShift nodes: %w", err)
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	nodes := make([]string, 0, len(lines))

	for _, line := range lines {
		if line == "" {
			continue
		}
		nodeName := strings.TrimPrefix(line, "node/")
		if nodeName != "" {
			nodes = append(nodes, nodeName)
		}
	}

	return nodes, nil
}

// logStats writes statistics to log files
func logStats(timestamp, statsType string, stats map[string]int) error {
	for statName, value := range stats {
		logFile := filepath.Join(logsDir, fmt.Sprintf("%s_%s.log", statsType, statName))
		f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			return err
		}
		fmt.Fprintf(f, "%s %d\n", timestamp, value)
		f.Close()
	}
	return nil
}

// runMonitor runs the main monitoring loop
func runMonitor() error {
	if err := setupDirectories(outputDir); err != nil {
		return err
	}

	log.Println("Starting EIP monitoring...")

	nodes, err := getNodes()
	if err != nil {
		return err
	}
	log.Printf("Found EIP-enabled nodes: %v", nodes)

	for {
		timestamp := time.Now().Format(time.RFC3339)

		// Get global statistics
		eipStats, err := getEIPStats()
		if err != nil {
			return err
		}

		cpicStats, err := getCPICStats()
		if err != nil {
			return err
		}

		// Log global statistics
		if err := logStats(timestamp, "ocp_eips", eipStats); err != nil {
			return err
		}
		if err := logStats(timestamp, "ocp_cpic", cpicStats); err != nil {
			return err
		}

		// Monitor individual nodes
		for _, node := range nodes {
			// Get node-specific EIP stats
			eipData, err := runOCCommand([]string{"get", "eip", "-o", "json"})
			if err != nil {
				log.Printf("Error monitoring node %s: %v", node, err)
				continue
			}

			eipAssigned := 0
			items, ok := eipData["items"].([]interface{})
			if ok {
				for _, item := range items {
					itemMap, ok := item.(map[string]interface{})
					if !ok {
						continue
					}

					status, ok := itemMap["status"].(map[string]interface{})
					if !ok {
						continue
					}

					statusItems, ok := status["items"].([]interface{})
					if !ok {
						continue
					}

					for _, statusItem := range statusItems {
						statusItemMap, ok := statusItem.(map[string]interface{})
						if !ok {
							continue
						}

						if statusItemMap["node"] == node {
							eipAssigned++
						}
					}
				}
			}

			// Get node-specific CPIC stats
			cpicData, err := runOCCommand([]string{"get", "cloudprivateipconfig", "-o", "json"})
			if err != nil {
				log.Printf("Error monitoring node %s: %v", node, err)
				continue
			}

			cpicSuccess := 0
			items, ok = cpicData["items"].([]interface{})
			if ok {
				for _, item := range items {
					itemMap, ok := item.(map[string]interface{})
					if !ok {
						continue
					}

					spec, ok := itemMap["spec"].(map[string]interface{})
					if !ok || spec["node"] != node {
						continue
					}

					status, ok := itemMap["status"].(map[string]interface{})
					if !ok {
						continue
					}

					conditions, ok := status["conditions"].([]interface{})
					if !ok {
						continue
					}

					for _, cond := range conditions {
						condMap, ok := cond.(map[string]interface{})
						if !ok {
							continue
						}

						if reason, ok := condMap["reason"].(string); ok && reason == "CloudResponseSuccess" {
							cpicSuccess++
							break
						}
					}
				}
			}

			// Log node statistics
			if err := logStats(timestamp, fmt.Sprintf("%s_ocp_eip", node), map[string]int{"assigned": eipAssigned}); err != nil {
				log.Printf("Error logging stats for node %s: %v", node, err)
			}
			if err := logStats(timestamp, fmt.Sprintf("%s_ocp_cpic", node), map[string]int{"success": cpicSuccess}); err != nil {
				log.Printf("Error logging stats for node %s: %v", node, err)
			}

			log.Printf("%s - EIP: %d, CPIC: %d", node, eipAssigned, cpicSuccess)
		}

		// Check if monitoring should continue
		if eipStats["assigned"] == eipStats["configured"] && cpicStats["success"] == eipStats["configured"] {
			break
		}

		time.Sleep(1 * time.Second)
	}

	log.Println("Monitoring complete - all EIPs assigned and CPIC issues resolved")
	return nil
}

// mergeLogs merges log files into data files
func runMergeLogs() error {
	if err := setupDirectories(outputDir); err != nil {
		return err
	}

	log.Println("Merging log files...")

	// Get all log files
	logFiles, err := filepath.Glob(filepath.Join(logsDir, "*.log"))
	if err != nil {
		return err
	}

	if len(logFiles) == 0 {
		return &EIPToolkitError{Message: "No log files found"}
	}

	// Get unique node names
	nodeSet := make(map[string]bool)
	for _, logFile := range logFiles {
		filename := filepath.Base(logFile)
		if !strings.HasPrefix(filename, "ocp_") && !strings.HasPrefix(filename, "azure_") {
			if idx := strings.Index(filename, "_"); idx > 0 {
				nodeName := filename[:idx]
				nodeSet[nodeName] = true
			}
		}
	}

	nodes := make([]string, 0, len(nodeSet))
	for node := range nodeSet {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	// Process each file type
	fileMappings := map[string]string{
		"ocp_cpic_success.log": "ocp_cpic_success.dat",
		"ocp_cpic_pending.log": "ocp_cpic_pending.dat",
		"ocp_cpic_error.log":   "ocp_cpic_error.dat",
		"ocp_eip_assigned.log": "ocp_eip_assigned.dat",
	}

	for logSuffix, datFilename := range fileMappings {
		dataFile := filepath.Join(dataDir, datFilename)
		outFile, err := os.Create(dataFile)
		if err != nil {
			return err
		}

		for _, node := range nodes {
			logFile := filepath.Join(logsDir, fmt.Sprintf("%s_%s", node, logSuffix))
			if _, err := os.Stat(logFile); err == nil {
				fmt.Fprintf(outFile, "\"%s\"\n", node)

				inFile, err := os.Open(logFile)
				if err != nil {
					outFile.Close()
					return err
				}

				buf := make([]byte, 1024)
				for {
					n, err := inFile.Read(buf)
					if n > 0 {
						outFile.Write(buf[:n])
					}
					if err != nil {
						break
					}
				}
				inFile.Close()
				fmt.Fprintf(outFile, "\n\n")
			}
		}
		outFile.Close()
	}

	log.Println("Log merge completed")
	return nil
}

// createPlots generates plots from data files (placeholder - requires external plotting tool)
func runCreatePlots() error {
	if err := setupDirectories(outputDir); err != nil {
		return err
	}

	log.Println("Creating plots...")

	dataFiles, err := filepath.Glob(filepath.Join(dataDir, "*.dat"))
	if err != nil {
		return err
	}

	if len(dataFiles) == 0 {
		return &EIPToolkitError{Message: "No .dat files found for plotting"}
	}

	log.Printf("Found %d data files. Plotting requires external tools (gnuplot, matplotlib, etc.)", len(dataFiles))
	log.Println("Plot generation completed")
	return nil
}

func main() {
	var rootCmd = &cobra.Command{
		Use:   "eip-toolkit-simple",
		Short: "EIP Toolkit - Simple Version",
	}

	var monitorCmd = &cobra.Command{
		Use:   "monitor",
		Short: "Monitor EIP and CPIC status",
		RunE: func(cmd *cobra.Command, args []string) error {
			if outputDir == "" {
				timestamp := time.Now().Format("060102_150405")
				outputDir = filepath.Join("..", "runs", timestamp)
			}
			return runMonitor()
		},
	}

	var mergeCmd = &cobra.Command{
		Use:   "merge [directory]",
		Short: "Merge log files into data files",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			outputDir = args[0]
			return runMergeLogs()
		},
	}

	var plotCmd = &cobra.Command{
		Use:   "plot [directory]",
		Short: "Generate plots from data files",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			outputDir = args[0]
			return runCreatePlots()
		},
	}

	var allCmd = &cobra.Command{
		Use:   "all",
		Short: "Run complete pipeline: monitor → merge → plot",
		RunE: func(cmd *cobra.Command, args []string) error {
			if outputDir == "" {
				timestamp := time.Now().Format("060102_150405")
				outputDir = filepath.Join("..", "runs", timestamp)
			}

			fmt.Println("🚀 Starting Complete EIP Pipeline")
			fmt.Println("📊 Phase 1: Monitoring...")
			if err := runMonitor(); err != nil {
				return err
			}
			fmt.Println("✅ Phase 1 Complete")

			fmt.Println("🔄 Phase 2: Merging logs...")
			if err := runMergeLogs(); err != nil {
				return err
			}
			fmt.Println("✅ Phase 2 Complete")

			fmt.Println("📈 Phase 3: Creating plots...")
			if err := runCreatePlots(); err != nil {
				return err
			}
			fmt.Println("✅ Phase 3 Complete")

			fmt.Println("🎉 PIPELINE COMPLETE!")
			fmt.Printf("📁 Outputs saved in: %s\n", outputDir)
			return nil
		},
	}

	rootCmd.AddCommand(monitorCmd, mergeCmd, plotCmd, allCmd)

	if err := rootCmd.Execute(); err != nil {
		if _, ok := err.(*EIPToolkitError); ok {
			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		} else {
			fmt.Fprintf(os.Stderr, "Unexpected error: %v\n", err)
		}
		os.Exit(1)
	}
}
